torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1517,17 +1534,33 @@ void ProcessGroupNCCL::workCleanupLoop() {
           // rank
           abort();
         }
+
         // Report desync state in case of timeout
         if (desyncDebug_ && timedOut) {
           try {
-            // Set shutdown mode, so the heartbeat monitor thread will not abort
-            // process immediately.
+            // Set shutdown mode, so the heartbeat monitor thread will not
+            // abort process immediately.
             collectiveDebugInfoMode_.store(true);
+            // Store debug info to storage. (By default to local disk)
+            auto dumpingDebugInfo = checkAndModifyCheckDumpingDebugInfo();
+            if (dumpingDebugInfo) {
+              debugInfoStoreThread =
+                  std::thread(&ProcessGroupNCCL::dumpDebuggingInfo, this);
+            }
             auto desyncMsg = getNCCLWatchdogDebugInfo();
             LOG(ERROR) << desyncMsg;
+            if (dumpingDebugInfo && !debugInfoStoreThread.joinable()) {
+              std::this_thread::sleep_for(
+                  std::chrono::milliseconds(kWatchdogThreadSleepMillis * 20));
+            }
+            // At this point, we either have already waited for
+            // `kWatchdogThreadSleepMillis * 20` or the thread has finished so
+            // that we mark the thread detach and the dump of debug info
+            // becomes "best effort".
+            debugInfoStoreThread.detach();
           } catch (const std::exception& e) {
             LOG(ERROR) << "Failed to retrieve NCCL_DESYNC_DEBUG report. "
-                       << " Please file an issue. Error: " << e.what();
+                        << " Please file an issue. Error: " << e.what();
```

Comment:
curious: does linter require that we must have this extra space?

Reply:
I think this is an error brought in accidentally and it is fixed and updated already.